File: clients/react-app/src/renderer/src/logic/transaction.js
import * as mercury_wasm from 'mercury-wasm';

const newTransaction = async(coin, toAddress, isWithdrawal, qtBackupTx, block_height, network) => {

    let coin_nonce = mercury_wasm.createAndCommitNonces(coin);

    let server_pubnonce = await window.api.signFirst(coin_nonce.sign_first_request_payload);

    if (server_pubnonce.startsWith("0x")) {
        server_pubnonce = server_pubnonce.substring(2);
    }

    coin.secret_nonce = coin_nonce.secret_nonce;
    coin.public_nonce = coin_nonce.public_nonce;
    coin.server_public_nonce = server_pubnonce;
    coin.blinding_factor = coin_nonce.blinding_factor;

    const serverInfo = await window.api.infoConfig();

    let new_block_height = 0;
    if (block_height == null) {
        let block_header = await window.api.electrumRequest({
            method: 'blockchain.headers.subscribe',
            params: []
        });
        new_block_height = block_header.height;
    } else {
        new_block_height = block_height;
    }

    const initlock = serverInfo.initlock;
    const interval = serverInfo.interval;
    const feeRateSatsPerByte = serverInfo.fee_rate_sats_per_byte;   

    let partialSigRequest = mercury_wasm.getPartialSigRequest(
        coin,
        new_block_height,
        initlock, 
        interval,
        feeRateSatsPerByte,
        qtBackupTx,
        toAddress,
        network,
        isWithdrawal);

    const serverPartialSigRequest = partialSigRequest.partial_signature_request_payload;

    let serverPartialSig = await window.api.signSecond(serverPartialSigRequest);

    if (serverPartialSig.startsWith("0x")) {
        serverPartialSig = serverPartialSig.substring(2);
    }

    const clientPartialSig = partialSigRequest.client_partial_sig;
    const msg = partialSigRequest.msg;
    const session = partialSigRequest.encoded_session;
    const outputPubkey = partialSigRequest.output_pubkey;

    const signature = mercury_wasm.createSignature(msg, clientPartialSig, serverPartialSig, session, outputPubkey);

    const encodedUnsignedTx = partialSigRequest.encoded_unsigned_tx;

    const signed_tx = mercury_wasm.newBackupTransaction(encodedUnsignedTx, signature);

    return signed_tx;
}

export default { newTransaction };

File: clients/apps/kotlin/src/main/kotlin/Transaction.kt
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*

class Transaction() {

    companion object {

        private suspend fun signFirst(clientConfig: ClientConfig, signFirstRequestPayload: SignFirstRequestPayload): String {

            val url = clientConfig.statechainEntity + "/" + "sign/first"

            val httpClient = HttpClient(CIO) {
                install(ContentNegotiation) {
                    json()
                }
            }

            val signFirstResponsePayload: SignFirstResponsePayload = httpClient.post(url) {
                contentType(ContentType.Application.Json)
                setBody(signFirstRequestPayload)
            }.body()

            httpClient.close()

            var serverPubNonceHex = signFirstResponsePayload.serverPubnonce

            if (serverPubNonceHex.startsWith("0x")) {
                serverPubNonceHex = serverPubNonceHex.substring(2)
            }

            return serverPubNonceHex
        }

        private suspend fun signSecond(clientConfig: ClientConfig, partialSigRequest: PartialSignatureRequestPayload) : String {

            val url = clientConfig.statechainEntity + "/" + "sign/second"

            val httpClient = HttpClient(CIO) {
                install(ContentNegotiation) {
                    json()
                }
            }

            val partialSignatureResponsePayload: PartialSignatureResponsePayload = httpClient.post(url) {
                contentType(ContentType.Application.Json)
                setBody(partialSigRequest)
            }.body()

            httpClient.close()

            var partialSigHex = partialSignatureResponsePayload.partialSig

            if (partialSigHex.startsWith("0x")) {
                partialSigHex = partialSigHex.substring(2)
            }

            return partialSigHex
        }

        suspend fun create(
            coin: Coin,
            clientConfig: ClientConfig,
            blockHeight: UInt?,
            qtBackupTx: UInt,
            toAddress: String,
            network: String,
            isWithdrawal: Boolean,
            feeRate: UInt?
        ) : String {
            val coinNonce = createAndCommitNonces(coin)
            coin.secretNonce = coinNonce.secretNonce
            coin.publicNonce = coinNonce.publicNonce
            coin.blindingFactor = coinNonce.blindingFactor

            coin.serverPublicNonce = signFirst(clientConfig, coinNonce.signFirstRequestPayload)

            var newBlockHeight: UInt = 0u

            if (blockHeight == null) {
                val electrumClient = getElectrumClient(clientConfig)

                val blockHeader = electrumClient.blockchainHeadersSubscribe()
                newBlockHeight = blockHeader.height.toUInt();

                electrumClient.closeConnection()
            } else {
                newBlockHeight = blockHeight
            }

            val infoConfig = getInfoConfig(clientConfig)

            val initlock = infoConfig.initlock;
            val interval = infoConfig.interval;

            val feeRateSatsPerByte = feeRate ?: infoConfig.feeRateSatsPerByte.toUInt()

            val partialSigRequest = getPartialSigRequest(
                coin,
                newBlockHeight,
                initlock,
                interval,
                feeRateSatsPerByte,
                qtBackupTx,
                toAddress,
                network,
                isWithdrawal)

            val serverPartialSigRequest = partialSigRequest.partialSignatureRequestPayload

            val serverPartialSig = signSecond(clientConfig, serverPartialSigRequest)

            val clientPartialSig = partialSigRequest.clientPartialSig
            val msg = partialSigRequest.msg
            val session = partialSigRequest.encodedSession
            val outputPubkey = partialSigRequest.outputPubkey

            val signature = createSignature(msg, clientPartialSig, serverPartialSig, session, outputPubkey)

            val encodedUnsignedTx = partialSigRequest.encodedUnsignedTx

            val signedTx = newBackupTransaction(encodedUnsignedTx, signature)

            return signedTx
        }
    }

}